src/protocol/codecs/msgpack.rs

```diff
@@ -0,0 +1,278 @@
+use std::{
+    fmt::{self, Debug, Display, Formatter},
+    str,
+};
+
+use data_encoding::{BASE32_NOPAD, BASE64};
+use serde::{
+    de::{Error, Visitor},
+    Deserialize, Deserializer, Serialize, Serializer,
+};
+
+type Period = u64;
+type Round = u64;
+type Step = u64;
+
+// A Seed holds the entropy needed to generate cryptographic keys.
+type Seed = Ed25519Seed;
+
+#[allow(unused)]
+type VrfProof = [u8; 80];
+
+/* Classical signatures */
+#[allow(unused)]
+type Ed25519Signature = [u8; 64];
+type Ed25519PublicKey = [u8; 32];
+#[allow(unused)]
+type Ed25519PrivateKey = [u8; 64];
+type Ed25519Seed = [u8; 32];
+
+/// A [ProposalValue] is a triplet of a block hashes (the contents themselves and the encoding of the block),
+/// its proposer, and the period in which it was proposed.
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct ProposalValue {
+    #[serde(default, rename = "oper")]
+    original_period: Period,
+
+    #[serde(rename = "oprop")]
+    original_proposer: HashDigest,
+
+    #[serde(rename = "dig")]
+    block_digest: HashDigest,
+
+    #[serde(rename = "encdig")]
+    encoding_digest: HashDigest,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct RawVote {
+    #[serde(rename = "snd")]
+    sender_addr: HashDigest,
+
+    #[serde(rename = "rnd")]
+    round: Round,
+
+    #[serde(default, rename = "per")]
+    period: Period,
+
+    #[serde(default, rename = "step")]
+    step: Step,
+
+    #[serde(default, rename = "prop")]
+    proposal: Option<ProposalValue>,
+}
+
+/// A OneTimeSignature is a cryptographic signature that is produced a limited
+/// number of times and provides forward integrity.
+///
+/// Specifically, a OneTimeSignature is generated from an ephemeral secret. After
+/// some number of messages is signed under a given OneTimeSignatureIdentifier
+/// identifier, the corresponding secret is deleted. This prevents the
+/// secret-holder from signing a contradictory message in the future in the event
+/// of a secret-key compromise.
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct OneTimeSignature {
+    // Sig is a signature of msg under the key PK.
+    //#[serde(rename = "s", deserialize_with = "deserialize_byte64_arr_opt")]
```

Comment:
Is this not needed? Is it worth documenting context for why these aren't currently in use?

Reply:
Aaaa, I should update the PR info with this:
I still haven't found a simple way to deserialize arrays larger than 32 bytes (and we still don't need that info for the basic test). But I do want to keep all the message info here, although we didn't parse/deserialize it yet as a placeholder and an indication that there is more stuff in these messages.

I can leave a file comment and also move there the small TODO from line 82:
```
// TODO(Rqnsom): deserialize 64 byte arrays.
```

Reply:
Done